reolink_aio/typings.py

```diff
@@ -27,6 +30,177 @@
     },
 )
 
+GetTimeDst = TypedDict(
+    "GetTimeDst",
+    {
+        "enable": bool,
+        "offset": int,
+    }
+)
+
+GetTimeDstRule = TypedDict(
+    "GetTimeDstRule",
+    {
+        "Mon": int,
+        "Week": int,
+        "Weekday": int,
+        "Hour": int,
+        "Min": int,
+        "Sec": int,
+    }
+)
+
+GetTime = TypedDict(
+    "GetTime",
+    {
+        "year": int,
+        "mon": int,
+        "day": int,
+        "hour": int,
+        "min": int,
+        "sec": int,
+        "hourFmt": int,
+        "timeFmt": str,
+        "timeZone": int,
+    }
+)
+
+GetTimeResponse = TypedDict(
+    "GetTimeResponse",
+    {
+        "Dst": GetTimeDst,
+        "Time": GetTime,
+    }
+)
+
+class _DST_Rule:
+
+    def __init__(self, data:dict[str,Any], prefix:str) -> None:
+        self.data:GetTimeDstRule = data
+        self.prefix = prefix
+
+    @property
+    def month(self)->int:
+        """rule month"""
+        return self.data[f"{self.prefix}Mon"]
+    
+    @property
+    def week(self)->int:
+        """rule week"""
+        return self.data[f"{self.prefix}Week"]
+
+    @property
+    def weekday(self)->int:
+        """rule weekday"""
+        # reolink sun=0, python mon = 0, sun = 6
+        return (self.data[f"{self.prefix}Weekday"] - 1) % 7
+    
+    @property
+    def hour(self)->int:
+        """rule hour"""
+        return self.data[f"{self.prefix}Hour"]
+
+    @property
+    def minute(self)->int:
+        """rule minute"""
+        return self.data[f"{self.prefix}Min"]
+
+    @property
+    def second(self)->int:
+        """rule second"""
+        return self.data[f"{self.prefix}Sec"]
+
+    def time(self)->time:
+        """rule as time"""
+        return time(self.hour, self.minute, self.second)
+    
+    def date(self, year:int)->date:
+        """rule as date with year"""
+        weekday =  self.weekday
+        # 5 is the last week of the month, not necessarily the 5th week
+        if self.week == 5:
+            if self.month < 12:
+                month = self.month + 1
+            else:
+                month = 1
+                year += 1
+            __date = date(year, month, 1)
+            if __date.weekday() < weekday:
+                __date -= timedelta(weeks=1)
+            __date -= timedelta(days=__date.weekday() - weekday)
+        else:
+            __date = date(year, self.month, 1) + timedelta(weeks=self.week)
+            __date += timedelta(days=__date.weekday() - weekday)
+        return __date
+
+    def datetime(self, year:int)->datetime:
+        """rule as datetime with year"""
+        return datetime.combine(self.date(year), self.time())
+
+class _YearStartStopCache(dict[str,tuple[datetime,datetime]]):
```

Comment:
I dont really understand what this does, but I would prefer to not create to many seperate classes.
Can we make this part of the Reolink_timezone class?

Reply:
I come from the school of logical separation of concerns, so I tend to break things out into their own domains. I can refactor it into a single function, and simple cache, but it felt like it was "obfuscating" what it was doing to much, so I built it split out to make things more readable.